allennlp_models/vision/dataset_readers/flickr30k.py

```diff
@@ -136,13 +144,55 @@ def __init__(
             self.model.eval()
             self.tokenizer = transformers.AutoTokenizer.from_pretrained("bert-large-uncased")
 
+        # feature cache
+        self.hard_negative_features_cache_dir = feature_cache_dir
+        self.hard_negative_coordinates_cache_dir = feature_cache_dir
+        self._hard_negative_features_cache_instance: Optional[MutableMapping[str, Tensor]] = None
+        self._hard_negative_coordinates_cache_instance: Optional[MutableMapping[str, Tensor]] = None
+        if self.hard_negative_features_cache_dir and self.hard_negative_coordinates_cache_dir:
+            logger.info(f"Calculating hard negatives with a cache at {self.feature_cache_dir}")
+
+    @property
+    def _hard_negative_features_cache(self) -> MutableMapping[str, Tensor]:
+        if self._hard_negative_features_cache_instance is None:
+            if self.hard_negative_features_cache_dir is None:
+                logger.info("could not find feature cache dir")
+                self._hard_negative_features_cache_instance = {}
+            else:
+                logger.info("found feature cache dir")
+                os.makedirs(self.feature_cache_dir, exist_ok=True)  # type: ignore
+                self._hard_negative_features_cache_instance = TensorCache(
+                    os.path.join(self.feature_cache_dir, "hard_negative_features"),  # type: ignore
+                    read_only=not self.write_to_cache,
+                )
+
+        return self._hard_negative_features_cache_instance
+
+    @property
+    def _hard_negative_coordinates_cache(self) -> MutableMapping[str, Tensor]:
+        if self._hard_negative_coordinates_cache_instance is None:
+            if self.hard_negative_coordinates_cache_dir is None:
+                self._hard_negative_coordinates_cache_instance = {}
+            else:
+                os.makedirs(self.feature_cache_dir, exist_ok=True)  # type: ignore
+                self._hard_negative_coordinates_cache_instance = TensorCache(
+                    os.path.join(self.feature_cache_dir, "hard_negative_coordinates"),  # type: ignore
+                    read_only=not self.write_to_cache,
+                )
+
+        return self._hard_negative_coordinates_cache_instance
+
     @overrides
     def _read(self, file_path: str):
         file_path = cached_path(file_path, extract_archive=True)
         files_in_split = set()
+        i = 0
         with open(file_path, "r") as f:
             for line in f:
+                if self.max_instances is not None and i >= self.max_instances:
+                    break
                 files_in_split.add(line.rstrip("\n"))
+                i += 5
```

Comment:
You can instead write something like this:
```Python
for i, line in enumerate(f):
    ...
```
... and then compare to `I*5`. Then you don't need to maintain that counter variable.

Reply:
Thanks!